src/animation.rs

```diff
@@ -0,0 +1,503 @@
+//!
+//! The created actions are enabled by default in the 'play' state. This means that
+//! when calling [`Mixer::update`] the created actions will begin to be played back
+//! immediately.
+//!
+//! ```rust,no_run
+//! # let mut window = three::Window::builder("", "").build();
+//! # let mut mixer = three::animation::Mixer::new();
+//! # let (group, _, _, clips) = window.factory.load_gltf("AnimatedScene.gltf");
+//! # window.scene.add(&group);
+//! let actions: Vec<three::animation::Action> = clips
+//!     .into_iter()
+//!     .map(|clip| mixer.action(clip))
+//!     .collect();
+//! ```
+//!
+//! ### Playing the animation back
+//!
+//! Finally, we run the animation actions by updating their [`Mixer`] in the main
+//! game loop.
+//!
+//! ```rust,no_run
+//! # let mut window = three::Window::builder("", "").build();
+//! # let camera = unimplemented!();
+//! # let mut mixer = three::animation::Mixer::new();
+//! # let (group, _, _, clips) = window.factory.load_gltf("AnimatedScene.gltf");
+//! # window.scene.add(&group);
+//! # let actions: Vec<three::animation::Action> = clips
+//! #     .into_iter()
+//! #     .map(|clip| mixer.action(clip))
+//! #    .collect();
+//! while window.update() {
+//!     mixer.update(window.input.delta_time());
+//!     window.render(&camera);
+//! }
+//! ```
+//!
+//! ### Putting it all together
+//!
+//! See the `gltf-animation` example for the full code.
+//!
+//! [`disable`]: struct.Action.html#method.disable
+//! [`play`]: struct.Action.html#method.play
+//! [`pause`]: struct.Action.html#method.pause
+//!
+//! [`Action`]: struct.Action.html
+//! [`Clip`]: struct.Clip.html
+//! [`Mixer`]: struct.Mixer.html
+//! [`Mixer::action`]: struct.Mixer.html#method.action
+//! [`Mixer::update`]: struct.Mixer.html#method.update
+
+use cgmath;
+use froggy;
+use mint;
+use std::sync::mpsc;
+
+use Object;
+
+/// Describes the interpolation behaviour between keyframes.
+#[derive(Clone, Copy, Debug, Eq, Hash, PartialEq)]
+pub enum Interpolation {
+    /// Immediate change between keyframe values.
+    Discrete,
+
+    /// Linear interpolation between keyframe values.
+    Linear,
+
+    /// Smooth cubic interpolation between keyframe values.
+    Cubic,
+
+    /// Smooth Catmull–Rom spline interpolation between keyframe values.
+    CatmullRom,
+}
+
+/// Describes the looping behaviour of an [`Action`].
+///
+/// [`Action`]: struct.Action.html
+#[derive(Clone, Copy, Debug, Eq, Hash, PartialEq)]
+pub enum LoopMode {
+    /// Play the clip exactly once, i.e. do not loop at all.
+    Once,
+
+    /// Play the clip with the chosen number of repetitions.
+    Repeat(Repetitions),
+
+    /// Play the clip with the chosen number of repetitions, alternatively
+    /// playing in forward and reverse order.
+    PingPong(Repetitions),
+}
+
+/// Describes the number of times an [`Action`] repeats itself.
+///
+/// [`Action`]: struct.Action.html
+#[derive(Clone, Copy, Debug, Eq, Hash, PartialEq)]
+pub enum Repetitions {
```

Comment:
not 100% we need a new type for that, as opposed to having something like 
```rust
Repeat { repetitions: Option<usize> },
```

Reply:
I prefer the other way as the intention is more clear. Also, this pattern is recommended by the [API guidelines](https://rust-lang-nursery.github.io/api-guidelines/type-safety.html#arguments-convey-meaning-through-types-not-bool-or-option-c-custom-type).

Reply:
I don't see that API guidelines section applicable here. We aren't talking about function arguments, where `bool` indeed becomes ambiguous for the lost semantics. In our case, there is always `repetitions` field involved, meaning the semantics is clear.

Reply:
If we call the field `max_repetitions: Option<usize>` then it would make more sense.

Reply:
I called it `limit` because that's shorter.